utils/date.h

```diff
@@ -3665,15 +3665,15 @@ class time_of_day_storage<std::chrono::duration<Rep, Period>, detail::classify::
         os << t.h_.count();
         switch (t.mode_)
         {
-        case time_of_day_storage::is24hr:
-            os << "00";
-            break;
         case am:
             os << "am";
             break;
         case pm:
             os << "pm";
             break;
+        default:
+            os << "00";
+            break;
```

Comment:
I'd consider leaving the is24hr case in place
and call on_internal_error_noexcept in the default case so we can catch and fix the undefined path in testing.
while having a safe fallback in production.

Reply:
Not using a default case is a deliberate decision: it allows the compiler to emit an error when a new member is added to the enum.
The only case such a switch can hit UB is if there is a memory corruption.

Reply:
Moreoever, in what sense is the default (and **unexpected**) case "undefined" in this code? It just doesn't print words like "am"/"pm"/"00", but it doesn't seem to invoke any "undefined behavior"... What am I missing?

Reply:
Thank you all for your feedback and suggestions. Should I throw an error in the default case? If the community prefers not to use a default case, I will respect that decision and close this PR.

Reply:
Our policy in general is to not implement the default case and use enum classes at the same time.  That lets the compiler warn us if an enum value is added and there's no explicitl case for it in the switch.
In cases like this where the switch is over an integer value that is nit an enum class, the compiler can't really help us at compile time, ao instead we can implement a default case for the unexpected values and do as I suggested above https://github.com/scylladb/scylladb/pull/18813#discussion_r1609597495, to help us catch the bug in testing and do something sensible and harmless in production.

Reply:
I see that the on_internal_error_noexcept method requires a logger object, but I haven't found a suitable logger object. How should I select one?

Reply:
You can add a `logging::logger pp_logger("pretty_printers");` in utils/pretty_printers.cc and use that.

Reply:
> I see that the on_internal_error_noexcept method requires a logger object, but I haven't found a suitable logger object. How should I select one?

First, this was a bad design. In Scylla I added a common `on_internal_error()` which doesn't need to take a logger name and uses some global one, exactly for this reason - see https://github.com/scylladb/scylladb/commit/259811b6ec8b03bcb37adb04c645522d2d0dc6d4. Ironically, because of this, Seastar hardly uses  the `on_internal_error()` API. The few places that does use it (actually on_fatal_internal_error() in all places..) use the global "seastar_logger". Just do "extern logger seastar_logger;" to access it, as several header files already do.

Second, I'm still not sure I understand why you want to change this code at all... Again, I don't understand where any "undefined behavior" was involved (and if there are none, then the commit message is wrong, and must be changed). It looks to me like this am/pm string will be missing when the enum is incorrect, but who cares?

I guess we can add a #ifdef SEASTAR_DEBUG and assert so that if there is a corruption in a date variable and it's detected during printing in a debugging run, we'll crash immediately. But the usefulness of this "protection" is rather dubious, if you ask me. If you already have memory corruption, you will need to be very lucky for it to happen inside some data variable which is going to be printed and not in a million other memory locations that will cause much more serious problems.

Reply:
> Second, I'm still not sure I understand why you want to change this code at all... Again, I don't understand where any "undefined behavior" was involved (and if there are none, then the commit message is wrong, and must be changed). It looks to me like this am/pm string will be missing when the enum is incorrect, but who cares?

Sorry for the misunderstanding in my description. I added a default case for adherence to coding standards, and I have already updated the description accordingly.